pyro/infer/contract.py

```diff
@@ -95,22 +95,29 @@ def dims(self, term):
         return [d for d in range(-term.dim(), 0) if term.size(d) > 1]
 
     def sumproduct(self, terms, dims):
-        assert all(dim < 0 for dim in dims)
-        if not dims:
-            return sum(terms)
-        shape = list(broadcast_shape(*set(x.shape for x in terms)))
-        for dim in dims:
-            shape[dim] = 1
-        shape.reverse()
-        while shape and shape[-1] == 1:
-            shape.pop()
-        shape.reverse()
-        shape = tuple(shape)
-        return logsumproductexp(terms, shape)
+        key = 'sumproduct', frozenset(id(x) for x in terms), frozenset(dims)
+        if key in self._cache:
+            return self._cache[key]
+
+        if dims:
+            assert all(dim < 0 for dim in dims)
+            shape = list(broadcast_shape(*set(x.shape for x in terms)))
+            for dim in dims:
+                shape[dim] = 1
+            term = logsumproductexp(terms, tuple(shape))
+        else:
+            term = sum(terms)
+
+        # Aggressively squeeze to improve sharing.
+        while term.dim() and term.size(0) == 1:
+            term = term.squeeze(0)
+        self._save_tensor(term)
+        self._cache[key] = term
+        return term
 
     def product(self, term, ordinal):
         for frame in sorted(ordinal, key=lambda f: -f.dim):
-            if term.shape[frame.dim] != 1:
+            if -frame.dim <= term.dim() and term.size(frame.dim) != 1:
```

Comment:
Any reason why this is changed to `.size` now?

Reply:
#1406

Reply:
Interesting! We changed from `.size` to `.shape` in the last release, but I suppose that was only for the sake of consistency, rather than perf.